File: agents/python_sdk/clove_sdk/mixins/state.py
"""State store syscalls.

Provides key-value storage: store, fetch, delete, list keys.
"""

from typing import Any, Optional, TYPE_CHECKING

from ..protocol import SyscallOp
from ..models import StoreResult, FetchResult, DeleteResult, KeysResult

if TYPE_CHECKING:
    from ..transport import Transport


class StateMixin:
    """Mixin for state store operations.

    Requires _transport attribute of type Transport.
    """

    _transport: 'Transport'

    def store(
        self,
        key: str,
        value: Any,
        scope: str = "global",
        ttl: Optional[int] = None
    ) -> StoreResult:
        """Store a key-value pair in the shared state store.

        Args:
            key: Storage key
            value: Value to store (must be JSON-serializable)
            scope: Storage scope - "global", "agent", or "session"
            ttl: Time-to-live in seconds (optional)

        Returns:
            StoreResult with success status
        """
        payload = {
            "key": key,
            "value": value,
            "scope": scope
        }
        if ttl is not None:
            payload["ttl"] = ttl

        result = self._transport.call_json(SyscallOp.SYS_STORE, payload)

        return StoreResult(
            success=result.get("success", False),
            error=result.get("error")
        )

    def fetch(self, key: str) -> FetchResult:
        """Fetch a value from the shared state store.

        Args:
            key: Storage key to fetch

        Returns:
            FetchResult with value if found
        """
        result = self._transport.call_json(
            SyscallOp.SYS_FETCH,
            {"key": key}
        )

        # Kernel returns "exists", SDK model uses "found"
        found = result.get("exists", result.get("found", False))

        return FetchResult(
            success=result.get("success", False),
            value=result.get("value"),
            found=found,
            error=result.get("error")
        )

    def delete_key(self, key: str) -> DeleteResult:
        """Delete a key from the shared state store.

        Args:
            key: Storage key to delete

        Returns:
            DeleteResult with deletion status
        """
        result = self._transport.call_json(
            SyscallOp.SYS_DELETE,
            {"key": key}
        )

        return DeleteResult(
            success=result.get("success", False),
            deleted=result.get("deleted", False),
            error=result.get("error")
        )

    def list_keys(self, prefix: str = "") -> KeysResult:
        """List keys in the shared state store.

        Args:
            prefix: Optional prefix to filter keys

        Returns:
            KeysResult with list of matching keys
        """
        payload = {"prefix": prefix} if prefix else {}

        result = self._transport.call_json(SyscallOp.SYS_KEYS, payload)

        return KeysResult(
            success=result.get("success", False),
            keys=result.get("keys", []),
            count=result.get("count", 0),
            error=result.get("error")
        )

File: sdk/typescript/src/api/state.ts
/**
 * State Store API.
 *
 * Provides key-value storage operations.
 */

import { SyscallOp } from "../protocol.js";
import { Transport } from "../transport.js";
import { StoreResult, FetchResult, DeleteResult, KeysResult } from "../models.js";

export type StateScope = "agent" | "global" | "world";

export interface StoreOptions {
  key: string;
  value: unknown;
  scope?: StateScope;
  ttl?: number; // TTL in seconds
}

export class StateAPI {
  constructor(private transport: Transport) {}

  /**
   * Store a key-value pair.
   */
  async store(options: StoreOptions): Promise<StoreResult> {
    const payload: Record<string, unknown> = {
      key: options.key,
      value: options.value,
      scope: options.scope ?? "agent",
    };

    if (options.ttl !== undefined) {
      payload.ttl = options.ttl;
    }

    const result = await this.transport.callJson(SyscallOp.SYS_STORE, payload);

    return {
      success: (result.success as boolean) ?? false,
      error: result.error as string | undefined,
    };
  }

  /**
   * Retrieve a value by key.
   */
  async fetch(key: string, scope: StateScope = "agent"): Promise<FetchResult> {
    const result = await this.transport.callJson(SyscallOp.SYS_FETCH, { key, scope });

    return {
      success: (result.success as boolean) ?? false,
      value: result.value,
      found: (result.found as boolean) ?? false,
      error: result.error as string | undefined,
    };
  }

  /**
   * Delete a key.
   */
  async delete(key: string, scope: StateScope = "agent"): Promise<DeleteResult> {
    const result = await this.transport.callJson(SyscallOp.SYS_DELETE, { key, scope });

    return {
      success: (result.success as boolean) ?? false,
      deleted: (result.deleted as boolean) ?? false,
      error: result.error as string | undefined,
    };
  }

  /**
   * List keys with optional prefix.
   */
  async keys(prefix?: string, scope: StateScope = "agent"): Promise<KeysResult> {
    const payload: Record<string, unknown> = { scope };
    if (prefix !== undefined) {
      payload.prefix = prefix;
    }

    const result = await this.transport.callJson(SyscallOp.SYS_KEYS, payload);

    return {
      success: (result.success as boolean) ?? false,
      keys: (result.keys as string[]) ?? [],
      count: (result.count as number) ?? 0,
      error: result.error as string | undefined,
    };
  }

  // Convenience methods

  /**
   * Get a value (returns undefined if not found).
   */
  async get<T = unknown>(key: string, scope: StateScope = "agent"): Promise<T | undefined> {
    const result = await this.fetch(key, scope);
    return result.found ? (result.value as T) : undefined;
  }

  /**
   * Set a value.
   */
  async set(key: string, value: unknown, scope: StateScope = "agent"): Promise<boolean> {
    const result = await this.store({ key, value, scope });
    return result.success;
  }
}
